sympy/tensor/array/tests/test_array_comprehension.py

```diff
@@ -0,0 +1,27 @@
+from sympy.tensor.array.array_comprehension import ArrayComprehension
+from sympy.tensor.array import ImmutableDenseNDimArray
+from sympy.abc import i, j, k, l
+from sympy.utilities.pytest import raises
+
+
+def test_array_comprehension():
+    a = ArrayComprehension(i*j, (i, 1, 3), (j, 2, 4))
+    b = ArrayComprehension(i, (i, 1, j+1))
+    c = ArrayComprehension(i+j+k+l, (i, 1, 2), (j, 1, 3), (k, 1, 4), (l, 1, 5))
+    d = ArrayComprehension(k, (i, 1, 5))
+    assert a.doit().tolist() == [[2, 3, 4], [4, 6, 8], [6, 9, 12]]
+    assert isinstance(b.doit(), ArrayComprehension)
+    assert isinstance(a.doit(), ImmutableDenseNDimArray)
+    assert b.subs(j, 3) == ArrayComprehension(i, (i, 1, 4))
+    assert c.free_symbols == {i, j, k, l}
```

Comment:
`c.free_symbols == set([])`. There are no free symbols in `c` as they are variables to be expanded.

You should subtract the iteration variables from `.expr.free_symbols`, then add the free symbols in the bounds.

Reply:
For example:

```python
>>> ArrayComprehension(a*i, (i, k, N)).free_symbols
{a, k, N}
```
where `i` is removed because it won't exist anymore after calling `.doit()`.

Reply:
Will this implementation allow multivariate lists like,
```
>>> A = ArrayComprehension(i + j, (i, ki, Ni), (j, kj, Nj))
>>> A.free_symbols
{kj, Nj, ki, Ni}
```

Reply:
Yes,  only `i` and `j` will be removed from free symbols in that example.